compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -1101,20 +1101,88 @@ private STNode parseFuncDefOrFuncTypeDesc(STNode metadata, STNode visibilityQual
+                                              List<STNode> qualifiers, STNode functionKeyword, boolean isObjectMember,
+                                              boolean isObjectTypeDesc) {
+        switch (peek(2).kind) {
+            case OPEN_PAREN_TOKEN:
                 STNode name = parseFunctionName();
-                STNode resourcePath = parseOptionalRelativePath(isObjectMember);
+                STNode resourcePath = STNodeFactory.createEmptyNodeList();
                 switchContext(ParserRuleContext.FUNC_DEF);
                 STNode funcSignature = parseFuncSignature(false);
                 STNode funcDef = parseFuncDefOrMethodDeclEnd(metadata, visibilityQualifier, qualifiers, functionKeyword,
                         name, resourcePath, funcSignature, isObjectMember, isObjectTypeDesc);
                 endContext();
                 return funcDef;
+            case DOT_TOKEN:
+            case IDENTIFIER_TOKEN:
+            case OPEN_BRACKET_TOKEN:
+                name = parseFunctionName();
+                resourcePath = parseRelativeResourcePath();
+                if (!isObjectMember) {
+                    addInvalidNodeToNextToken(resourcePath,
+                                              DiagnosticErrorCode.ERROR_RESOURCE_PATH_IN_FUNCTION_DEFINITION);
+                    resourcePath = STNodeFactory.createEmptyNodeList();
+                }
+                switchContext(ParserRuleContext.FUNC_DEF);
+                funcSignature = parseFuncSignature(false);
+                funcDef = parseFuncDefOrMethodDeclEnd(metadata, visibilityQualifier, qualifiers, functionKeyword,
+                        name, resourcePath, funcSignature, isObjectMember, isObjectTypeDesc);
+                endContext();
+                return funcDef;
+            case EQUAL_TOKEN:
+            case SEMICOLON_TOKEN:
+                endContext();
+                return parseVarDeclWithFunctionType(metadata, visibilityQualifier, qualifiers, functionKeyword,
+                                                    STNodeFactory.createEmptyNode(), isObjectMember, isObjectTypeDesc,
+                                                    false);
+            default:
+                STToken functionName = consume();
+                STToken token = peek();
+                recover(token, ParserRuleContext.OPTIONAL_RELATIVE_PATH, metadata, visibilityQualifier, qualifiers,
+                        functionKeyword, isObjectMember, isObjectTypeDesc);
+                resourcePath = parseOptionalRelativePath(isObjectMember);
+                switchContext(ParserRuleContext.FUNC_DEF);
+                funcSignature = parseFuncSignature(false);
+                funcDef = parseFuncDefOrMethodDeclEnd(metadata, visibilityQualifier, qualifiers, functionKeyword,
+                        functionName, resourcePath, funcSignature, isObjectMember, isObjectTypeDesc);
+                endContext();
+                return funcDef;
```

Comment:
We need to add recovery for semicolon and equal as well. Use a new parser rule context. Using `OPTIONAL_RELATIVE_PATH` for recovery is wrong here. In the parser after recovery, we always make a recursive call except for `KEEP` action

Reply:
Also, we don't need to pass arguments for the current method to recovery. since it is outdated logic.